tools/api_proto_plugin/type_context.py

```diff
@@ -0,0 +1,199 @@
+"""Type context for FileDescriptorProto traversal."""
+
+from collections import namedtuple
+
+from tools.api_proto_plugin import annotations
+
+# A comment is a (raw comment, annotation map) pair.
+Comment = namedtuple('Comment', ['raw', 'annotations'])
+
+
+class SourceCodeInfo(object):
+  """Wrapper for SourceCodeInfo proto."""
+
+  def __init__(self, name, source_code_info):
+    self.name = name
+    self.proto = source_code_info
+    # Map from path to SourceCodeInfo.Location
+    self._locations = {
+        str(location.path): location for location in self.proto.location
+    }
+    self._file_level_comments = None
+    self._file_level_annotations = None
+
+  @property
+  def file_level_comments(self):
+    """Obtain inferred file level comment."""
+    if self._file_level_comments:
+      return self._file_level_comments
+    comments = []
+    earliest_detached_comment = max(
+        max(location.span) for location in self.proto.location)
```

Comment:
I think this finds the last detached comment?

That said, I think location.span is an array of either `[start_line, start_col, end_line, end_col]` or `[start_line, start_col, end_col]`. Seems like you'd want to just use `location.span[0]` instead of `max(location.span)` to avoid accidentally choosing a column number as the max. And use <= for comparisons to the `earliest_detached_comment` below.

Reply:
I see this was just lifted from the old implementation. Still seems wrong, but I suppose changing it might alter the output.

Reply:
It actually finds the last in the loop below. I will fix to use just `span[0]` and verify output, thanks.